### PR TITLE
LongRunning OBO cache fix for different user assertions

### PR DESCRIPTION
**Pull request by trwalke**
Fixes #
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3825

**Changes proposed in this request**
Enabling obo request to acquire new token if user cached assertion is different from the one provided.
This is done by ensuring `InitiateLongRunningProcessInWebApi` always goes to get a new token rather than using what is already in the cache for a specified cache key.

**Testing**
Unit

**Performance impact**

**Documentation**
- [ ] All relevant documentation is updated.

### CODE REVIEW

**Comment by trwalke** [comment]
Seems like it makes sense to check if the cached assertion is the same when using `InitiateLongRunningProcessInWebApi`.
initialize indicates that you want to start a new long running session and the previous one should not interfere even though the same session key is used. `AcquireTokenInLongRunningProcess` should still respect the cached user assertion however. 

Thoughts @bgavrilMS @pmaytak

**Comment by pmaytak** [comment]
My initial thought - your conclusion is right. If cache key exists, scopes are the same, but assertion is different, we should replace the cached ATs and RTs. Currently we'll return the cached AT if it's not expired. [Here's](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/16b744a4c56efb20073693c081a3512aa385da25/tests/Microsoft.Identity.Test.Unit/PublicApiTests/OBOTests.cs#L229) the test for these scenarios. So we'd need to update it with an extra Initiate call.

Would this have any behavioral breaking changes? Currently if users call Initiate for existing key, they'd expect a cached token; but with the change, each new user assertion would result in a network call, which makes sense - if new user assertion, then give my an updated token.

I'll have to think a bit more about this.

File: src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs [hunk 1]
```diff
@@ -97,7 +99,7 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
             // No AT in the cache or AT needs to be refreshed
             try
             {
-                if (cachedAccessToken == null)
+                if (!isCachedAccessTokenValid)
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs:102]: This won't exactly work. RefreshRt method would try to use the cached RT to get a token and AAD will throw. So need to make sure to just call get new AT method right away.

File: src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs [hunk 2]
```diff
@@ -55,7 +56,8 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
                     cachedAccessToken = await CacheManager.FindAccessTokenAsync().ConfigureAwait(false);
                 }
 
-                if (cachedAccessToken != null)
+                isCachedAccessTokenValid = cachedAccessToken != null && IsCachedAssertionTheSame(cachedAccessToken.Secret);
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs:59]: Thinking if this filtering could be done in FindAT/FindRT methods in the TokenCache class, but it's probably more efficient to do it here.

Comment (pmaytak) [reply src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs:59]: After discussion - the filtering should probably be done here. Also a method like IsInitiateLROBO called would also be helpful.

**Comment by pmaytak** [comment]
This test is one that would need to be updated.
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/665971b6d514c9e2e3cefc90e9ae3cbcaa06550f/tests/Microsoft.Identity.Test.Unit/PublicApiTests/OBOTests.cs#L262

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/OBOTests.cs [hunk 3]
```diff
@@ -225,6 +225,51 @@ public async Task AcquireTokenByObo_InLongRunningProcess_TestAsync()
...
+
+                Assert.AreEqual(TestConstants.ATSecret2, result.AccessToken);
+                Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+                cachedAccessToken = cca.UserTokenCacheInternal.Accessor.GetAllAccessTokens().Single();
+                cachedRefreshToken = cca.UserTokenCacheInternal.Accessor.GetAllRefreshTokens().Single();
+                Assert.AreEqual(oboCacheKey, cachedAccessToken.OboCacheKey);
+                Assert.AreEqual(oboCacheKey, cachedRefreshToken.OboCacheKey);
+                Assert.AreEqual(TestConstants.ATSecret2, cachedAccessToken.Secret);
+                userCacheAccess.AssertAccessCounts(0, 2);
+            }
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/OBOTests.cs:270]: I'd also add another test case here for same cache key and scopes but different assertion:
```csharp
result = await cca.InitiateLongRunningProcessInWebApi(TestConstants.s_scope, $"{TestConstants.DefaultAccessToken}2", ref oboCacheKey)
```

File: src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs [hunk 4]
```diff
@@ -42,6 +42,15 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
             AuthenticationResult authResult = null;
 
             CacheRefreshReason cacheInfoTelemetry = CacheRefreshReason.NotApplicable;
+
+            //Check if initiating a long running process
+            if (AuthenticationRequestParameters.UserAssertion != null && !string.IsNullOrEmpty(AuthenticationRequestParameters.LongRunningOboCacheKey))
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs:47]: nit: have a little helper like `IsLongOBOInitialize` kind of helper.

**Review by bgavrilMS — APPROVED** [review]

**Review by pmaytak — APPROVED** [review]

**Pull request merged**